src/join/composer.jl

```diff
@@ -88,238 +91,11 @@ _rename_cols(old_names::AbstractVector{Symbol},
-        disallow_sorted |= rcat ⊻ lcat
-    end
-
-    # TODO:
-    # If DataAPI.invrefpool vectors are found in the "on" columns
-    # then potentially the following optimizations can be done:
-    # 1. identify rows in shorter table that should be dropped
-    # 2. develop custom _innerjoin_sorted and _innerjoin_unsorted that
-    #    drop rows from shorter table that do not match rows from longer table based on
-    #    PooledArray refpool check
-    # This optimization significantly complicates the code (especially sorted path).
-    # It should be added if in practice we find that the use case is often enough
-    # and that the benefits are significant. The two cases when the benefits should
-    # be expected are:
-    # 1. Shorter table is sorted when we drop rows not matching longer table rows
-    # 2. Shorter table does not have duplicates when we drop rows not matching longer table rows
-
-    left_col = prepare_on_col(left_cols...)
-    right_col = prepare_on_col(right_cols...)
-
-    local left_ixs
-    local right_ixs
-
-    if isempty(left_col) || isempty(right_col)
-        # we treat this case separately so we know we have at least one element later
-        left_ixs, right_ixs = Int[], Int[]
-    else
-        # if sorting is not disallowed try using a fast algorithm that works
-        # on sorted columns; if it is not run or errors fall back to the unsorted case
-        # the try-catch is used to handle the case when columns on which we join
-        # contain values that are not comparable
-        already_joined = false
-        if !disallow_sorted
-            try
-                if issorted(left_col) && issorted(right_col)
-                    left_ixs, right_ixs = _innerjoin_sorted(left_col, right_col)
-                    already_joined = true
-                end
-            catch
-                # nothing to do - one of the columns is not sortable
-            end
-        end
-        if !already_joined
-            if right_shorter
-                if left_col isa AbstractVector{<:Union{Integer, Missing}} &&
-                   right_col isa AbstractVector{<:Union{Integer, Missing}}
-                    left_ixs, right_ixs = _innerjoin_unsorted_int(left_col, right_col)
-                else
-                    left_ixs, right_ixs = _innerjoin_unsorted(left_col, right_col)
-                end
-            else
-                if left_col isa AbstractVector{<:Union{Integer, Missing}} &&
-                   right_col isa AbstractVector{<:Union{Integer, Missing}}
-                    right_ixs, left_ixs = _innerjoin_unsorted_int(right_col, left_col)
-                else
-                    right_ixs, left_ixs = _innerjoin_unsorted(right_col, left_col)
-                end
-            end
-        end
-    end
-
+    left_ixs, right_ixs = find_inner_rows(joiner)
     dfl = joiner.dfl[left_ixs, :]
     dfr_noon = joiner.dfr[right_ixs, Not(joiner.right_on)]
```

Comment:
@nalimilan - and these lines are crucial for `innerjoin`. Note that here I am using `getindex` not `copyto!` which leads to different performance characteristics (in different cases we are slow/fast).

We should make a decision if we should use `copyto!` everywhere (or maybe we can ensure that the user can ignore the choice and just use what is natural to use - as in the current proposal)